src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -86,14 +94,69 @@ private BoundExpression RewriteCollectionExpressionConversion(Conversion convers
+                LocalRewriter self,
+                SyntheticBoundNodeFactory factory,
+                TypeWithAnnotations listElementType,
+                [NotNullWhen(true)] out BoundExpression? result)
             {
-                if (!ConversionsBase.IsSpanOrListType(compilation, node.Type, WellKnownType.System_Collections_Generic_List_T, out elementType))
+                result = null;
+
+                if (node.Elements is not [BoundCollectionExpressionSpreadElement singleSpread])
+                {
+                    return false;
+                }
+
+                var iEnumerableOfTType = compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+                var iEnumerableOfElementType = iEnumerableOfTType.Construct([listElementType]);
+
+                var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+                var singleSpreadExpression = singleSpread.Expression;
+                var conversion = compilation.Conversions.ClassifyImplicitConversionFromExpression(singleSpreadExpression, iEnumerableOfElementType, ref discardedUseSiteInfo);
+                if (!conversion.IsIdentity && !conversion.IsReference)
                 {
                     return false;
                 }
+
+                if (singleSpread.EnumeratorInfoOpt?.GetEnumeratorInfo.Method.ReturnType.IsValueType == true)
+                {
+                    var iCollectionOfTType = compilation.GetSpecialType(SpecialType.System_Collections_Generic_ICollection_T);
+                    var iCollectionOfElementType = iCollectionOfTType.Construct([listElementType]);
+
+                    var singleSpreadType = singleSpreadExpression.Type;
+                    Debug.Assert(singleSpreadType is not null);
+
+                    if (!singleSpreadType.ImplementsInterface(iCollectionOfElementType, ref discardedUseSiteInfo))
+                    {
+                        return false;
+                    }
+                }
+
+                if (self.TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Linq_Enumerable__ToList, out MethodSymbol? toListGeneric, isOptional: true))
+                {
+                    var toListOfElementType = toListGeneric.Construct([listElementType]);
+                    result = factory.Call(receiver: null, toListOfElementType, singleSpreadExpression);
+                    return true;
+                }
+                else if (self.TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Collections_Generic_List_T__ctorIEnumerable, out MethodSymbol? listCtorEnumerableGeneric, isOptional: true))
+                {
+                    var listCtorEnumerableOfElementType = listCtorEnumerableGeneric.AsMember((NamedTypeSymbol)node.Type);
+                    result = new BoundObjectCreationExpression(node.Syntax, listCtorEnumerableOfElementType, singleSpreadExpression);
+                    return true;
+                }
```

Comment:
[`ToList` uses `new List<T>(expr)` constructor under the hood](https://github.com/dotnet/runtime/blob/21fe57174af159f7bcb39f591cd9fbb96a36bfcb/src/libraries/System.Linq/src/System/Linq/ToCollection.cs#L60-L75), bud adds an additional Linq-related optimization on top. Therefore we should prefer `ToList` over List constructor since both have equal IL size. As I stated in the PR description, we can get even smarter and emit List constructor when we know at compile time, that the spreading collection cannot be a Linq enumerable, but I didn't implement that part in this PR.

> Do we need to optimize the case when `Enumerable.ToList<T>()` is unavailable?

`Enumerable.ToList<T>()` lives in `System.Linq.dll`, while `new List<T>(expr)` in the `System.Runtime.dll`. I can imagine a constrained environment (e.g. some IoT app built with NativeAOT), in which `System.Linq.dll` isn't even included into compilation, so I think, having a fallback optimization makes sense. Plus the maintenance cost of keeping this optimization is relatively low (1 well-known member, which is probably not gonna be touched ever again, this small 5-lines-of-code branch, that does the work and a test)

Reply:
I don't think we should optimize code paths that are not going to be hit. Let's remove the `new List<T>(expr)` case. Thanks.

Reply:
Done, please take a look